nats/rtt_command.go

```diff
@@ -0,0 +1,197 @@
+}
+
+func (c *rttCmd) rtt(_ *kingpin.ParseContext) error {
+	targets, err := c.targets()
+	if err != nil {
+		return err
+	}
+
+	err = c.performTest(targets)
+	if err != nil {
+		return err
+	}
+
+	if c.json {
+		j, err := json.MarshalIndent(&targets, "", "  ")
+		if err != nil {
+			return err
+		}
+
+		fmt.Println(string(j))
+
+		return nil
+	}
+
+	f := fmt.Sprintf("%%%ds: %%v\n", c.calcIndent(targets, 3))
+
+	for _, t := range targets {
+		fmt.Printf("%s:\n\n", t.URL)
+
+		for _, r := range t.Results {
+			fmt.Printf(f, r.Address, r.RTT)
+		}
+
+		fmt.Println()
+	}
+
+	return nil
+}
+
+func (c *rttCmd) calcIndent(targets []*rttTarget, prefix int) int {
+	i := prefix
+
+	for _, t := range targets {
+		for _, r := range t.Results {
+			p := len(r.Address) + prefix
+
+			if p > i {
+				i = p
+			}
+		}
+	}
+
+	return i
+}
+
+func (c *rttCmd) performTest(targets []*rttTarget) (err error) {
+	for _, target := range targets {
+		opts := natsOpts()
+		if target.tlsName != "" {
+			opts = append(opts, func(o *nats.Options) error {
+				o.TLSConfig = &tls.Config{MinVersion: tls.VersionTLS12, ServerName: target.tlsName}
+				return nil
+			})
+		}
+
+		for _, r := range target.Results {
+			r.Time = time.Now()
+			r.URL, r.RTT, err = c.calcRTT(target.URL, opts)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (c *rttCmd) calcRTT(server string, opts []nats.Option) (string, time.Duration, error) {
+	nc, err := newNatsConn(server, opts...)
+	if err != nil {
+		return "", 0, err
+	}
+	defer nc.Close()
+
+	time.Sleep(25 * time.Millisecond)
+
+	var totalTime time.Duration
+	for i := 1; i <= c.iterations; i++ {
+		start := time.Now()
+		nc.Flush()
+		rtt := time.Since(start)
+		totalTime += rtt
+	}
+
+	return nc.ConnectedUrl(), totalTime / time.Duration(c.iterations), nil
+}
+
+func (c *rttCmd) targets() (targets []*rttTarget, err error) {
+	for _, s := range strings.Split(servers, ",") {
+		if !strings.Contains(s, "://") {
+			s = fmt.Sprintf("nats://%s", s)
+		}
+
+		u, err := url.Parse(s)
+		if err != nil {
+			return targets, err
+		}
+
+		targets = append(targets, &rttTarget{URL: u.String()})
+		target := targets[len(targets)-1]
+
+		// its a ip just add it
+		if net.ParseIP(u.Hostname()) != nil {
+			target.Results = append(target.Results, &rttResult{Address: u.Hostname()})
+			continue
+		}
+
+		// else look it up and add all its addresses
+		addrs, _ := net.LookupHost(u.Hostname())
+		if len(addrs) == 0 {
+			target.Results = append(target.Results, &rttResult{Address: u.Hostname()})
+			continue
+		}
+
+		// if we have many addresses we'll connect to each IP but we have to use the
+		// name of the original server address to do validate TLS, connect here, check it
+		// requires TLS and store the name to use when connecting to each IP
+		nc, err := newNatsConn(u.Hostname(), natsOpts()...)
+		if err != nil {
+			return targets, err
+		}
+
+		if nc.TLSRequired() {
```

Comment:
@derekcollison I got this from your PR but I think its redundant, cos it's always tlsname being u.Hostname() and if the server doesnt require TLS it will just be ignored, so this pre-connection isnt needed we can always set it.

What do you think? Sounds right?

Reply:
Since we are decomposing the A records for each target, which will be IPs, we need to know if the connection will be TLS, either because of us or the server. We need to place the option that will use the correct name when doing TLS.

Reply:
I undertand why we're doing it - I dont think we need to connect to nats to actually do it though is my point. 

Only reason to connect is to check if TLSRequired() and then we add the option. But we can just add it always since NATS client will ignore the option when the server isn't requiring TLS.  So there is no need to gate this around TLSRequired()

Reply:
Ah, if you have tested it and that works agree.

Reply:
OK, seems to work, was just wonering what I am missing, guess we'll find out :)

Reply:
For some reason I thought if you specified a TLS config in client it would fail against a non-TLS client.

Reply:
Me too :) however..here's a very mixed set of servers all in one go, seems solid

```
% nats rtt --creds ngs.creds -s connect.ngs.global,localhost,demo.nats.io:4443,demo.nats.io:4222
nats://connect.ngs.global:

       nats://35.246.211.176:: 42.466147ms
       nats://35.234.115.220:: 42.224727ms
       nats://35.234.118.122:: 43.293272ms

nats://localhost:

                nats://[::1]:: 148.805µs
            nats://127.0.0.1:: 126.379µs

nats://demo.nats.io:4443:

   nats://107.170.221.32:4443: 202.51536ms

nats://demo.nats.io:4222:

   nats://107.170.221.32:4222: 203.318011ms

```

Reply:
That is a good one to think about. If the client expects TLS I would want it to fail if the server is not TLS. /cc @kozlovic

Reply:
You are not setting any TLS option in your example. `-creds` is not setting `TLS Secure`, so your client is set for no TLS but will gladly try to connect to a TLS server.
The question is: should UserCredentials() set the internal Options.Secure = true? Is a server *forced* to use TLS in order to accept clients with `--creds`? I don't think so..

Reply:
sorry some context is missing here @kozlovic - I found that if I always set `o.TLSConfig = &tls.Config{MinVersion: tls.VersionTLS12, ServerName: target.tlsName}` even for non TLS servers, its always accepted and no errors.

Reply:
So your are not passing the TLS config through the nats.Secure() option setter? Are you using the "old way" of connecting from the option object Options.Connect() as opposed to nats.Connect()?
Yes, if for some reason the TLSConfig is provided "directly" (bypassing nats.Secure()), then the checkForSecure() is flawed since it only uses o.Secure boolean to know if the user wants secure. This boolean is set in several option setters that deal with TLS. It could be very easy to strengthen the checkForSecure() function by also checking that TLSConfig is not nil...

Reply:
I did my own Option since Secure() is not really secure ;)

Reply:
Does SkipVerify IIRC

Reply:
Yeh we may have to do that, I copied some code Derek had lying about where he very carefully only set it on servers that required it - and I noticed it wasnt needed since its not failing if I just always set it.

So if it's desired to fail when tlsconfig is set, we should check and fail

Reply:
The code should have been something like:
```
if target.tlsName != "" {
    opts = append(opts, nats.Secure(&tls.Config{MinVersion: tls.VersionTLS12, ServerName: target.tlsName}))
}
```
So do you want me to fix the Go client' checkForSecure() that it should also check for TLSConfig and not just the Secure boolean, or are you going to fix the way you add the option?

Reply:
@derekcollison No, it does not skip verify. The Secure() option does this:
```
func Secure(tls ...*tls.Config) Option {
	return func(o *Options) error {
		o.Secure = true
		// Use of variadic just simplifies testing scenarios. We only take the first one.
		if len(tls) > 1 {
			return ErrMultipleTLSConfigs
		}
		if len(tls) == 1 {
			o.TLSConfig = tls[0]
		}
		return nil
	}
}
```

Reply:
https://github.com/nats-io/jetstream/blob/916f8123401b58a6a24978eb562f2d93adaaf8c0/nats/rtt_command.go#L48-L57 original code fwiw, easier to grok whats done there.

If I just dont gate it around TLSRequired() and always set it, it never fails

Reply:
It's makeTLSConn that *used* to be not secure. It now does this:
```
func (nc *Conn) makeTLSConn() error {
	// Allow the user to configure their own tls.Config structure.
	var tlsCopy *tls.Config
	if nc.Opts.TLSConfig != nil {
		tlsCopy = util.CloneTLSConfig(nc.Opts.TLSConfig)
	} else {
		tlsCopy = &tls.Config{}
	}
...
```
and when creating the TLSConfig if not provided we *used* to skip verify, but we no longer do.

Reply:
ok so we just need to set the o.Secure to true if we wanted, but they way it is now is ok since we are using it just optionally when TLS is involved and multiple A records are in the hostname.